src/main/java/bio/terra/service/snapshotbuilder/SnapshotBuilderService.java

```diff
@@ -1,22 +1,155 @@
+  private final AzureSynapsePdao azureSynapsePdao;
+  private final AzureSynapseService azureSynapseService;
+  private final DatasetDao datasetDao;
 
-  public SnapshotBuilderService(SnapshotBuilderSettingsDao snapshotBuilderSettingsDao) {
+  public SnapshotBuilderService(
+      SnapshotBuilderSettingsDao snapshotBuilderSettingsDao,
+      AzureSynapsePdao azureSynapsePdao,
+      AzureSynapseService azureSynapseService,
+      DatasetDao datasetDao) {
     this.snapshotBuilderSettingsDao = snapshotBuilderSettingsDao;
+    this.azureSynapsePdao = azureSynapsePdao;
+    this.azureSynapseService = azureSynapseService;
+    this.datasetDao = datasetDao;
   }
 
-  public SnapshotBuilderSettings getSnapshotBuilderSettings(UUID datasetId) {
-    return snapshotBuilderSettingsDao.getSnapshotBuilderSettingsByDatasetId(datasetId);
+  private SnapshotBuilderDomainOption generateDataForDomain(SnapshotBuilderDomain domain) {
+    // TODO: Generate concept count and participant count.
+    return new SnapshotBuilderDomainOption()
+        .category(domain.getCategory())
+        .id(domain.getId())
+        .root(domain.getRoot());
+  }
+
+  private SnapshotBuilderSettingsOptions snapshotBuilderSettingsWithOptions(
+      UUID datasetId, SnapshotBuilderSettings settings, AuthenticatedUserRequest userRequest) {
+    return new SnapshotBuilderSettingsOptions()
+        .domainOptions(
+            settings.getSelectableDomains().stream().map(this::generateDataForDomain).toList())
+        .programDataOptions(
+            settings.getSelectableProgramData().stream()
+                .map(
+                    programData ->
+                        generateTableInformationForProgramDataOption(
+                            datasetId, programData, userRequest))
+                .toList())
+        .datasetConceptSets(settings.getPrepackagedDatasetConceptSets())
+        .featureValueGroups(settings.getFeatureValueGroups());
+  }
+
+  public SnapshotBuilderSettingsOptions getSnapshotBuilderSettings(
+      UUID datasetId, AuthenticatedUserRequest userRequest) {
+
+    SnapshotBuilderSettings snapshotBuilderSettings =
+        snapshotBuilderSettingsDao.getSnapshotBuilderSettingsByDatasetId(datasetId);
+
+    return snapshotBuilderSettingsWithOptions(datasetId, snapshotBuilderSettings, userRequest);
+  }
+
+  public SnapshotBuilderProgramDataOption generateTableInformationForProgramDataOption(
+      UUID datasetId,
+      SnapshotBuilderProgramData programData,
+      AuthenticatedUserRequest userRequest) {
+    return switch (programData.getKind()) {
+      case LIST -> new SnapshotBuilderProgramDataOption()
+          .id(programData.getId())
+          .name(programData.getName())
+          .kind(SnapshotBuilderProgramDataOption.KindEnum.RANGE)
+          .values(List.of());
+      case RANGE -> generateRangeInformation(datasetId, programData, userRequest);
+    };
+  }
+
+  public SnapshotBuilderProgramDataOption generateRangeInformation(
```

Comment:
This method has a lot of nesting and seems to repeat code across the different branches, as well as outside of the service.  I'm finding it a bit hard to follow.

Have you considered using existing method [DatasetService.retrieveColumnStatistics](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/dataset/DatasetService.java#L581) as the base of this method, or otherwise sharing code with it?

You then might be able to do something like this:
```
…
ColumnStatisticsModel columnStatistics = datasetService.retrieveColumnStatistics(userRequest, datasetId, "person", programDataMetadata.getColumnName(), "");
if (columnStatistics instanceof ColumnStatisticsDoubleModel columnStatisticsDouble) {
  return response.min(columnStatisticsDouble.getMinValue().intValue()).max(…);
} else if (columnStatistics instanceof ColumnStatisticsIntModel columnStatisticsInt) {
  return response.min(columnStatisticsInt.getMinValue()).max(…);
}
throw new BadRequestException(
              "Snapshot builder settings not configured correctly for dataset. Should be numeric data type");
```

Reply:
I thought about it - I like the approach you have shared though, will update.

Reply:
Closing the loop for posterity:

Ultimately this addition wasn't needed anyway, given a shift in approach to have the UI call the existing `getColumnStatistics` endpoint.  (Spiritually similar to this earlier suggestion though!)